Agents/FHAgent/FHAgent/src/main/java/uk/ac/cam/cares/jps/agent/fh/FHAgentDerivation.java

```diff
@@ -0,0 +1,278 @@
+                    RawToDerivedMap.put(mapping[0], mapping[1]);
+                    if(!iriMap.containsKey(mapping[0])){
+                        iriMap.put(mapping[0], ONTODERIV+mapping[0]+UUID.randomUUID());
+                    }
+                    
+                    if(!iriMap.containsKey(mapping[1])){
+                        iriMap.put(mapping[1], outputToIRIMap.get(mapping[1]));
+                    }
+                }
+            }
+            catch(Exception e){
+                throw new JPSRuntimeException("Failed to instantiate agent instances: " + e);
+            }
+    }
+
+    /*
+     * Save the IRI mapping to IRIMapper.txt
+     * @param mappingFile The location of IRIMapper.txt
+     */
+    private void saveIRIMapping (String mappingFile) throws IOException {
+
+        FileWriter fileWriter = new FileWriter(mappingFile);
+        PrintWriter printWriter = new PrintWriter(fileWriter);
+        
+        
+        for (String varName: iriMap.keySet()) {
+            printWriter.print(varName+"="+iriMap.get(varName)+"\r\n");
+        }
+
+        printWriter.close();
+        
+    }
+
+    /*
+     * Retrieve IRI map between variable key and IRI form IRIMapper.txt. 
+     * If the file does not exist yet, create new IRI and file instead.
+     * @param agentPropFile location of the agent.properties
+     * @param iriMappingFile location of the IRIMapper.txt
+     * @param outputToIRIMap a map between the variable key and IRI of the occupiedState produced by the agent
+     */
+    private void retrieveIRIMapping (String agentPropFile, String mappingFile, Map<String, String> outputToIRIMap){
+        File mapping = new File(mappingFile);
+        if(mapping.exists()) { 
+            try (FileReader fr = new FileReader(mapping);
+                BufferedReader br = new BufferedReader(fr);) {
+                String line;
+                while ((line = br.readLine()) != null) {
+                    String[] iriPair = line.split("=");
+                    iriMap.put(iriPair[0], iriPair[1]);
+                }
+            } catch (IOException e) {
+                throw new JPSRuntimeException("Failed to retrieve IRI map from mapping file: " + e);
+            }
+        }
+        else {
+            try{
+                createIRI(agentPropFile, outputToIRIMap);
+                saveIRIMapping(mappingFile);
+            }
+            catch (Exception e){
+                throw new JPSRuntimeException("Failed to create IRI mapping: " + e);
+            }
+            
+        }
+    }
+
+    /*
+     * Create agent and derivatiopn instances for all derivation pair
+     * @param agenURL the URL from which the agent is called
+     */
+    public void instantiateAgent(String agentURL) {
+        if (agentURL.equals(null)){
+            agentURL = derivationInstanceBaseURL;
+        }
+        try{
+            for(String rawVar : RawToDerivedMap.keySet()){
```

Comment:
Found cause of minor bug:

RawToDerivedMap is empty if IRIMapper.txt already exist
RawToDerivedMap is only populated under CreateIRI (Line 140)
CreateIRI is only executed (Line 218) when IRIMapper.txt does not exist

Reply:
Agent does not instantiate rdf:type om:Length and rdf:type ontodevice:OccupiedState for the variables. Might make more sense for the FH-Agent to do this? 

Agent's supposed workflow requires the agent to be used together with the DevInstAgent but this information was omitted from the README previously. Will need to work out what is the workflow of the FH-Agent with the DevInstAgent and include more information in the README.